src/rust-crypto/OutgoingRequestProcessor.ts

```diff
@@ -201,4 +204,84 @@ export class OutgoingRequestProcessor {
 
         return await this.http.authedRequest<string>(method, path, queryParams, body, opts);
     }
+
+    private async fetchWithRetry(
+        method: Method,
+        path: string,
+        queryParams: QueryDict,
+        body: string,
+        maxRetryCount: number = 3,
+    ): Promise<string> {
+        let currentRetryCount = 0;
+
+        // eslint-disable-next-line no-constant-condition
+        while (true) {
+            try {
+                return await this.rawJsonRequest(method, path, queryParams, body);
+            } catch (e) {
+                if (currentRetryCount >= maxRetryCount) {
+                    // Max number of retries reached, rethrow the error
+                    throw e;
+                }
+
+                currentRetryCount++;
+
+                const maybeRetryAfter = this.shouldWaitBeforeRetryingMillis(e);
```

Comment:
There is already a class that implements the retry logic.  It would probably be better to use that instead of writing something else here.  https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/scheduler.ts#L54

Reply:
Yes I have seen it. Indeed I think that there should be a retry mecanism inside the `FetchHttpApi` maybe with a new flag in `IRequestOpts` `retriable` (default true as most mx request are indempotent). 
Then the event scheduler will benefit from it too (unless there needs to be special logic for events). 

The thing is that it would impact the application a bit everywhere, I'd like to have support for the web team if we want to go that way.

(on side note I noticed that the scheduler has blocked retry on 502/M_TOO_LARGE, might not be common for rust outgoing request, but might want to exclude it too)

Reply:
There are a two different suggestions in this thread:

 * `MatrixScheduler` and `OutgoingRequestManager`  both implement logic to decide whether to retry a request, and if so when. We could extract that logic and share it, rather than having two copies. (Presumably, we would make a new method somewhere, and have `MatrixScheduler.RETRY_BACKOFF_RATELIMIT` call it.) That seems like quite a good idea to me. @BillCarsonFr: any reason we can't do this?
 * Move the retry mechanism down to `FetchHttpApi`, so that everything (including `MatrixScheduler` and `OutgoingRequestManager`) can benefit from it. This seems like a *much* more invasive change, and one I don't think we should do right now.

Reply:
> any reason we can't do this?

The `RETRY_BACKOFF_RATELIMIT` function is a bit strange, it has an used parameter, it is trying to cast a `MatrixError` into a `ConnectionError` (this can't work? or maybe at runtime?)
It doesn't want to retry on ConnectionError (like if the client internet connection is down), and we want that.

So I think we can't use it directly as it doesn't seem to be the same logic, and touching it will be also quite invasive. The only thing we could take easily is the exponential backoff, but it's just one line.

It's still bad that the logic doesn't match, but we might want to unify that in a second step?

Reply:
Ok, so after discussion I extracted a common retry alg between the scheduler and the outgoing request processor.
2 slight changes on the previous state:
 - Outgoing request processor try 5 times in total instead of 4
 - The matrix scheduer is not retrying on client timeouts anymore (as this timeout is already > 1mn depending on browser)
 
 Here is the refactoring commit https://github.com/matrix-org/matrix-js-sdk/pull/4061/commits/c2126a6f5d3960e467360bdc1a42ed629b1a3208